pkg/client/k8s/v1alpha1/resource.go

```diff
@@ -170,55 +170,86 @@ func (r *resource) List(opts metav1.ListOptions) (u *unstructured.UnstructuredLi
+// given resource
+type ResourceApplyType struct {
+	*resource
+
+	// various executors required to perform Apply
 	Getter  ResourceGetter
 	Creator ResourceCreator
 	Updater ResourceUpdater
+
+	// IsSkipUpdate will not update this resource if set to true.
+	// In other words, enabling this flag can only create the
+	// resource in the cluster if not created previously
+	IsSkipUpdate bool
 }
 
-// createOrUpdate is a resource that is suitable to be executed as an apply
-// operation
-type createOrUpdate struct {
-	*resource
-	options ResourceApplyOptions // options used during resource's apply operation
+// ResourceApplySkipUpdate sets IsSkipUpdate based
+// on the provided flag
+func ResourceApplySkipUpdate(skip bool) ResourceApplyTypeOption {
+	return func(r *ResourceApplyType) {
+		r.IsSkipUpdate = skip
+	}
 }
 
-// CreateOrUpdate returns a new instance of createOrUpdate resource
-func CreateOrUpdate(gvr schema.GroupVersionResource, namespace string) *createOrUpdate {
+// NewResourceApplier returns a new instance of ResourceApplyType
+func NewResourceApplier(
+	gvr schema.GroupVersionResource,
+	namespace string,
+	options ...ResourceApplyTypeOption,
+) *ResourceApplyType {
 	resource := Resource(gvr, namespace)
-	options := ResourceApplyOptions{Getter: resource, Creator: resource, Updater: resource}
-	return &createOrUpdate{resource: resource, options: options}
+	t := &ResourceApplyType{
+		resource: resource,
+		Getter:   resource,
+		Creator:  resource,
+		Updater:  resource,
+	}
+	for _, o := range options {
+		o(t)
+	}
+	return t
 }
 
 // Apply applies a resource to the kubernetes cluster. In other words, it
 // creates a new resource if it does not exist or updates the existing resource.
-func (r *createOrUpdate) Apply(obj *unstructured.Unstructured, subresources ...string) (resource *unstructured.Unstructured, err error) {
-	if r.options.Getter == nil {
+func (r *ResourceApplyType) Apply(
+	obj *unstructured.Unstructured,
+	subresources ...string,
+) (resource *unstructured.Unstructured, err error) {
+	if r.Getter == nil {
 		err = errors.New("nil resource getter instance: failed to apply resource")
 		return
 	}
-	if r.options.Creator == nil {
+	if r.Creator == nil {
 		err = errors.New("nil resource creator instance: failed to apply resource")
 		return
 	}
-	if r.options.Updater == nil {
+	if r.Updater == nil {
 		err = errors.New("nil resource updater instance: failed to apply resource")
 		return
 	}
 	if obj == nil {
 		err = errors.New("nil resource instance: failed to apply resource")
 		return
 	}
-	resource, err = r.options.Getter.Get(obj.GetName(), metav1.GetOptions{})
+	resource, err = r.Getter.Get(obj.GetName(), metav1.GetOptions{})
 	if err != nil {
 		if apierrors.IsNotFound(errors.Cause(err)) {
-			return r.options.Creator.Create(obj, subresources...)
+			return r.Creator.Create(obj, subresources...)
 		}
 		return nil, err
 	}
-	return r.options.Updater.Update(resource, obj, subresources...)
+	if r.IsSkipUpdate {
```

Comment:
Good to log when we are skipping the update

Reply:
Yes we can decide in caller also

Reply:
Done.. Added log inside Apply.